File: examples/java/Case05.java
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

final class Case05 {
    static final class Change {
        final String shipment;
        final int revision;
        final String state;

        Change(String shipment, int revision, String state) {
            this.shipment = shipment;
            this.revision = revision;
            this.state = state;
        }
    }

    static final class Projection {
        final Map<String, Change> records = new HashMap<>();

        synchronized Change read(String shipment) { return records.get(shipment); }
    }

    // faultscope:begin fs-c05.apply-on-arrival
    static void applyOnArrival(Projection projection, Change incoming) {
        synchronized (projection) {
            projection.records.put(incoming.shipment, incoming); // E42 can overwrite E43.
        }
    }
    // faultscope:end fs-c05.apply-on-arrival

    // faultscope:begin fs-c05.reject-stale-revision
    static boolean rejectStaleRevision(int applied, int incoming) {
        return incoming <= applied; // Equal revision is a duplicate.
    }
    // faultscope:end fs-c05.reject-stale-revision

    // faultscope:begin fs-c05.accept-newer-revision
    static void acceptNewerRevision(Projection projection, Change incoming) {
        projection.records.put(incoming.shipment, incoming); // State and revision together.
    }
    // faultscope:end fs-c05.accept-newer-revision

    // faultscope:begin fs-c05.apply-if-newer
    static boolean applyIfNewer(Projection projection, Change incoming) {
        synchronized (projection) { // Models one atomic projection mutation.
            Change current = projection.records.get(incoming.shipment);
            if (current != null && rejectStaleRevision(current.revision, incoming.revision)) {
                return false;
            }
            acceptNewerRevision(projection, incoming);
            return true;
        }
    }
    // faultscope:end fs-c05.apply-if-newer

    static void addShipmentTag(Set<String> tags, String tag) { tags.add(tag); }
}

File: examples/go/case05.go
package examplesgo

import "sync"

// ShipmentChange has ordering meaning only within one source-owned shipment stream.
type ShipmentChange struct {
	Shipment string
	Revision int
	State    string
}

type ShipmentProjection struct {
	mu      sync.Mutex
	records map[string]ShipmentChange
}

func NewShipmentProjection() *ShipmentProjection {
	return &ShipmentProjection{records: make(map[string]ShipmentChange)}
}

func (p *ShipmentProjection) Read(shipment string) ShipmentChange {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.records[shipment]
}

// faultscope:begin fs-c05.apply-on-arrival
func ApplyOnArrival(p *ShipmentProjection, incoming ShipmentChange) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.records[incoming.Shipment] = incoming // E42 can overwrite E43.
}

// faultscope:end fs-c05.apply-on-arrival

// faultscope:begin fs-c05.reject-stale-revision
func RejectStaleRevision(applied, incoming int) bool {
	return incoming <= applied // Equal revision is a duplicate.
}

// faultscope:end fs-c05.reject-stale-revision

// faultscope:begin fs-c05.accept-newer-revision
func AcceptNewerRevision(p *ShipmentProjection, incoming ShipmentChange) {
	p.records[incoming.Shipment] = incoming // State and revision change together.
}

// faultscope:end fs-c05.accept-newer-revision

// faultscope:begin fs-c05.apply-if-newer
func ApplyIfNewer(p *ShipmentProjection, incoming ShipmentChange) bool {
	p.mu.Lock() // Models one atomic projection mutation per entity.
	defer p.mu.Unlock()
	current := p.records[incoming.Shipment]
	if RejectStaleRevision(current.Revision, incoming.Revision) {
		return false
	}
	AcceptNewerRevision(p, incoming)
	return true
}

// faultscope:end fs-c05.apply-if-newer

// Set union is a negative control: order cannot change the final tag set.
func AddShipmentTag(tags map[string]bool, tag string) { tags[tag] = true }
